src/projects/bestfirst.clj

```diff
@@ -6,80 +6,49 @@
-    :liverpool ( {:location "warrington" :cost 2} {:location "manchester" :cost 7} {:location "chester" :cost 4})
-    })
-
-(defn bestFirstLMG [state]
-  "this function returns all possible routes that the user/machine can take from its current location."
-  (let [location (:state state)
-        cost (:cost state)]
-    (for [{loc :location cos :cost} (busRoutes (keyword location))]  (hash-map :cost (+ cos cost) :state loc))))
-
-(defn allStations []
-  "calling this function returns
-  the count of all stations which can be visited."
-  (count (set (flatten (for [x busRoutes] (for [{loc :location cos :cost} (second x) ] loc))))))
-
-(defn removeBeenValues [states toRemove]
-  "after using the legal move generator
-  you can pass the result to this function
-  to remove stations you have already visited."
-  (remove nil? (for [ {state :state cos :cost} states] (if (some #(= state %)toRemove) nil (hash-map :cost cos :state state)))))
-
-(defn fixPrice [currentLoc previousLoc]
-  "this function is used to update the
-  price if we have to back track to a previous
-  location we have already visited."
-  (first (remove nil? (for [{loc :location cost :cost} (busRoutes (keyword previousLoc))] (if (= loc currentLoc) cost )))))
+(use '[projects.helperFunctions :refer :all])
+(use '[projects.scenarios :refer :all])
 
 (defn best-firstSearchA
   "you can use this function to find a cheap route
   between two locations through many locations
   with different possible costs - using the best first search algorithm"
-  ([state goal lmg]
-   (best-firstSearchA state goal lmg []))
-  ([state goal lmg been]
+  ([state goal lmg map]
+   (best-firstSearchA state goal lmg map []))
+  ([state goal lmg map been]
    (cond
      (= (:state state) goal )
          (do (conj been (:state state)) state)
-     (= (count been) allStations)
+     (= (count (set been)) (allStations map))
          false
-     (empty? (removeBeenValues (lmg state) been))
+     (empty? (removeBeenValues (lmg map state) been))
          (do (println state " - FAIL")
-             (best-firstSearchA {:state (last been) :cost (- (state :cost) (fixPrice (:state state) (last been)))} goal lmg (conj been (:state state))))
+             (best-firstSearchA {:state (last been) :cost (- (state :cost) (fixPrice map (:state state) (last been)))} goal lmg map (conj been (:state state))))
      :else
      (do (println state)
-         (best-firstSearchA (first (sort-by :cost (removeBeenValues (lmg state) been))) goal lmg (conj been (:state state)))))))
+         (best-firstSearchA (first (sort-by :cost (removeBeenValues (lmg map state) been))) goal lmg map (conj been (:state state)))))))
 
-;;(best-firstSearchA {:state "newcastle" :cost 0} "chester" bestFirstLMG)
+;;(best-firstSearchA {:state "newcastle" :cost 0} "chester" bestFirstLMG busRoutes)
 ;;Find out Elapsed time for best-firstSearchA to run:
 ;;(time (best-firstSearchA {:state "newcastle" :cost 0} "chester" bestFirstLMG))
```

Comment:
change this to (time (best-firstSearchA {:state "newcastle" :cost 0} "chester" bestFirstLMG))